include/exadg/incompressible_navier_stokes/spatial_discretization/operator_coupled.cpp

```diff
@@ -1208,24 +1208,30 @@ OperatorCoupled<dim, Number>::apply_inverse_negative_laplace_operator(VectorType
   {
     // solve a linear system of equations for negative Laplace operator to given (relative)
     // tolerance using the PCG method
-    VectorType const * pointer_to_src = &src;
+    dst = 0.0;
     if(this->is_pressure_level_undefined())
     {
-      VectorType vector_zero_mean;
-      vector_zero_mean = src;
-
+      VectorType vector_zero_mean = src;
       if(laplace_operator->operator_is_singular())
       {
         dealii::VectorTools::subtract_mean_value(vector_zero_mean);
       }
-
-      pointer_to_src = &vector_zero_mean;
+      else
+      {
+        AssertThrow(false,
+                    dealii::ExcMessage(
+                      "Undefined pressure level always leads to singular Laplace operator."));
+      }
+      // Note that the preconditioner is not updated here since it has
+      // already been updated in the function update_block_preconditioner().
+      solver_pressure_block->solve(dst, vector_zero_mean);
+    }
+    else
+    {
+      // Note that the preconditioner is not updated here since it has
+      // already been updated in the function update_block_preconditioner().
+      solver_pressure_block->solve(dst, src);
     }
-
-    dst = 0.0;
-    // Note that the preconditioner is not updated here since it has
-    // already been updated in the function update_block_preconditioner().
-    solver_pressure_block->solve(dst, *pointer_to_src);
```

Comment:
`vector_zero_mean` goes out of scope?
i just replicated the functionality, but I think:
1) `dst=0` gets rid of the initial guess ... so is not really a good idea?
2) an undefined pressure level (no Neumann BC) in the Navier-Stokes problem always leads to a singular Laplace operator. So why the second if? Is there something I forget? How would the operator not be singular in that case? Is the wording of the functions misleading and I dont get it?

Reply:
Yes we are in the preconditioner, specifically `OperatorCoupledapply_preconditioner_pressure_block()` (see line 1119 above) . We solve the Schur complement `S p = b`. I think `dst` constains the last pressure iterate, which might help here, but I do not know. To verify, we would need to test. Should I add a 
`TODO: test if setting dst = 0; impacts iteration counts`
?
I think for now, let us reproduce the previous behavior, as we are here just for the compile warnings ([exactly what I did in the new version]).

Reply:
Regarding the compile warning and actual problem:
.) I am not at all against calling `solver->solve()` once, that would be beautiful, but it does not work in the present form, since for `this->is_pressure_level_undefined() = true`, we define `VectorType vector_zero_mean = src;` then `pointer_to_src = &vector_zero_mean;`, BUT the `vector_zero_mean` goes out of scope. So the pointer points to freed memory?

Now that you say it: if we really want to `dst=0;` (in the current implementation this is still optional), we can use `dst` as the temporary storage since it never goes out of scope. I can come up with a version that uses `dst` that way, you prefer that over the version that I did?

Reply:
Ah, thanks for the hint regarding the scope. I didn't see this. The declaration of the vector should happen outside the brackets.

Reply:
If I remember correctly, for some Krylov solvers, the preconditioner is applied to the residual, so the outcome of the preconditioner application would in general not be close to the solution of the linear system of equations?

Reply:
Overall, this else branch is not intended to be used in practise. For the Laplace operator, just the multigrid V-cycle should be a good preconditioner and nested Krylov solvers should generally be avoided. Or do you have an application where you see a major difference regarding this if/else-branch?

Reply:
Ah now I get it. You mean that the iterative solver should not be used to solve the pressure block in general in a production run:
```
if(this->param.exact_inversion_of_laplace_operator == false)
{
  // standard path
}
else
{
  // debug path, not to be used in production
}
```
That's true, yes :)

Reply:
> > If I remember correctly, for some Krylov solvers, the preconditioner is applied to the residual, so the outcome of the preconditioner application would in general not be close to the solution of the linear system of equations?
> 
> ![image](https://private-user-images.githubusercontent.com/69867655/372457350-5f1da9ce-6930-4c10-bb7b-9744d384993d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3ODkzOTMsIm5iZiI6MTcyNzc4OTA5MywicGF0aCI6Ii82OTg2NzY1NS8zNzI0NTczNTAtNWYxZGE5Y2UtNjkzMC00YzEwLWJiN2ItOTc0NGQzODQ5OTNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAxVDEzMjQ1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiYTJiOWFiZWUxZDU1YTk1MDZjNDM0NTVjMTlhYjg0YzkxOWU0YmI1ZjhlMWM2YWZjYzcxNTZmYTU4MzY2MmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yFSEwe4puvpK_CuPyTMu_tTM-IeEyAiAnNDo2e-HZD0)
> 
> Yes 👍 So `dst = 0;` makes a lot of sense now. we can use it as a temporary vector. Then we do not need an additional vector. I will adapt to the style you used before!

reply to myself: obviously one cant use `dst` as the temporary storage (like `dst = src; dst -=mean(dst);`) since one cant solve using a single vector as `dst` and `src` and the latter is `const`.